src/iceberg/update/update_schema.h

```diff
@@ -0,0 +1,451 @@
+
+/// \brief API for schema evolution.
+///
+/// When committing, these changes will be applied to the current table metadata.
+/// Commit conflicts will not be resolved and will result in a CommitFailed error.
+class ICEBERG_EXPORT UpdateSchema : public PendingUpdate {
+ public:
+  static Result<std::shared_ptr<UpdateSchema>> Make(
+      std::shared_ptr<Transaction> transaction);
+
+  ~UpdateSchema() override;
+
+  /// \brief Allow incompatible changes to the schema.
+  ///
+  /// Incompatible changes can cause failures when attempting to read older data files.
+  /// For example, adding a required column and attempting to read data files without
+  /// that column will cause a failure. However, if there are no data files that are
+  /// not compatible with the change, it can be allowed.
+  ///
+  /// This option allows incompatible changes to be made to a schema. This should be
+  /// used when the caller has validated that the change will not break. For example,
+  /// if a column is added as optional but always populated and data older than the
+  /// column addition has been deleted from the table, this can be used with
+  /// RequireColumn() to mark the column required.
+  ///
+  /// \return Reference to this for method chaining.
+  UpdateSchema& AllowIncompatibleChanges();
+
+  /// \brief Add a new optional top-level column.
+  ///
+  /// Because "." may be interpreted as a column path separator or may be used in
+  /// field names, it is not allowed in names passed to this method. To add to nested
+  /// structures or to add fields with names that contain ".", use AddColumn(parent,
+  /// name, type, ...).
+  ///
+  /// If type is a nested type, its field IDs are reassigned when added to the
+  /// existing schema.
+  ///
+  /// The added column will be optional with a null default value.
+  ///
+  /// \param name Name for the new column.
+  /// \param type Type for the new column.
+  /// \return Reference to this for method chaining.
+  /// \throws InvalidArgument If name contains ".".
+  UpdateSchema& AddColumn(std::string_view name, std::shared_ptr<Type> type);
+
+  /// \brief Add a new optional top-level column with documentation.
+  ///
+  /// Because "." may be interpreted as a column path separator or may be used in
+  /// field names, it is not allowed in names passed to this method. To add to nested
+  /// structures or to add fields with names that contain ".", use AddColumn(parent,
+  /// name, type, doc).
+  ///
+  /// If type is a nested type, its field IDs are reassigned when added to the
+  /// existing schema.
+  ///
+  /// The added column will be optional with a null default value.
+  ///
+  /// \param name Name for the new column.
+  /// \param type Type for the new column.
+  /// \param doc Documentation string for the new column.
+  /// \return Reference to this for method chaining.
+  /// \throws InvalidArgument If name contains ".".
+  UpdateSchema& AddColumn(std::string_view name, std::shared_ptr<Type> type,
```

Comment:
Should we unify this by removing the above one and adding a default parameter to `doc`?